# 1. Data Processing

## 1.1 Quarterly Reports Dataset:

After downloading the quarter reports financial tables from FACSET.COM , let's have a look at one example

In [13]:
import pandas as pd

# Example TESLA
file_path = 'Quarter Reports/TESLA.xlsx'

# Read the Excel file into a DataFrame
df = pd.read_excel(file_path)

# Print the first few rows of the DataFrame
print(df.head())


                                  Tesla, Inc. (TSLA) Unnamed: 1 Unnamed: 2  \
0                                            $163.57        NaN        NaN   
1                                                NaN        NaN        NaN   
2                                                NaN        NaN        NaN   
3                                        Tesla, Inc.        NaN        NaN   
4  TSLA   88160R101   B616C79   NASDAQ    Common ...        NaN        NaN   

  Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
0        NaN        NaN        NaN        NaN        NaN        NaN   
1        NaN        NaN        NaN        NaN        NaN        NaN   
2        NaN        NaN        NaN        NaN        NaN        NaN   
3        NaN        NaN        NaN        NaN        NaN        NaN   
4        NaN        NaN        NaN        NaN        NaN        NaN   

  Unnamed: 9  ... Unnamed: 51 Unnamed: 52 Unnamed: 53 Unnamed: 54 Unnamed: 55  \
0        NaN  ...      

The output shows the first five rows of a DataFrame from Tesla's quarterly report Excel file, consisting of 61 columns most of which are unnamed and filled with NaN values. The initial rows contain some metadata about Tesla like the stock price $163.57 ticker (TSLA) and information means that the main data needs cleaning for analysis. Therefore first we extract the relavant data from the quarter report tables 

In [14]:
import os
import openpyxl
import csv

def extract_data_from_xlsx(input_file, output_file):
    # Load the Excel file
    wb = openpyxl.load_workbook(input_file)
    
    # Assuming the data starts from row 9 and column 2
    sheet = wb.active
    max_row = sheet.max_row
    max_column = sheet.max_column
    
    # Open CSV file for writing
    with open(output_file, 'w', newline='') as csv_file:
        writer = csv.writer(csv_file)
        
        row_headers = [sheet.cell(row=row, column=1).value for row in range(9, max_row + 1)]
        
        writer.writerow([''] + row_headers)
        
        for col in range(2, max_column + 1):
            data_col = [sheet.cell(row=row, column=col).value for row in range(9, max_row + 1)]
            writer.writerow([sheet.cell(row=8, column=col).value] + data_col)

def extract_data_from_folder(input_folder, output_folder):
    # Create output folder if it doesn't exist
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)
    
    # Iterate over files in input folder
    for filename in os.listdir(input_folder):
        if filename.endswith(".xlsx"):
            input_file = os.path.join(input_folder, filename)
            output_file = os.path.join(output_folder, filename.replace(".xlsx", ".csv"))
            extract_data_from_xlsx(input_file, output_file)

# Start extracting
extract_data_from_folder("Quarter Reports", "Cleaned Quarter Reports")

#Example output
df = pd.read_csv("Cleaned Quarter Reports/TESLA.csv")
df

Unnamed: 0  Income Statement      Sales  Gross Income      EBIT    EBITDA  \
0     DEC '23               NaN  25167.000      4438.000  2064.000  3296.000   
1     SEP '23               NaN  23350.000      4178.000  1764.000  2999.000   
2     JUN '23               NaN  24927.000      4533.000  2399.000  3553.000   
3     MAR '23               NaN  23329.000      4511.000  2664.000  3710.000   
4     DEC '22               NaN  24318.000      5777.000  3935.000  4924.000   
5     SEP '22               NaN  21454.000      5382.000  3688.000  4644.000   
6     JUN '22               NaN  16934.000      4234.000  2606.000  3528.000   
7     MAR '22               NaN  18756.000      5460.000  3603.000  4483.000   
8     DEC '21               NaN  17719.000      4847.000  2613.000  3461.000   
9     SEP '21               NaN  13757.000      3660.000  2055.000  2816.000   
10    JUN '21               NaN  11958.000      2884.000  1335.000  2016.000   
11    MAR '21               NaN  10389.000      2215.000   493.000  1114.000   
12    DEC '20               NaN  10744.000      2066.000   575.000  1193.000   
13    SEP '20               NaN   8771.000      2063.000   809.000  1393.000   
14    JUN '20               NaN   6036.000      1267.000   327.000   894.000   
15    MAR '20               NaN   5985.000      1234.000   283.000   836.000   
16    DEC '19               NaN   7384.000      1391.000   347.000   924.000   
17    SEP '19               NaN   6303.000      1191.000   261.000   791.851   
18    JUN '19               NaN   6349.676       921.046   -50.113   528.459   
19    MAR '19               NaN   4541.464       565.743  -478.360   -10.783   
20    DEC '18               NaN   7225.873      1442.900   419.151   915.888   
21    SEP '18               NaN   6824.413      1528.265   442.941   742.941   
22    JUN '18               NaN   4002.231       625.930  -517.958  -266.158   
23    MAR '18               NaN   3408.751       462.326  -596.974  -180.741   
24    DEC '17               NaN   3288.249       438.786  -598.141  -128.535   
25    SEP '17               NaN   2984.675       459.940  -535.480  -134.856   
26    JUN '17               NaN   2789.557       674.015  -240.916   -64.316   
27    MAR '17               NaN   2696.270       674.046  -257.549   119.053   
28    DEC '16               NaN   2284.631       374.878  -327.098    -0.159   
29    SEP '16               NaN   2298.436       658.135    99.822   380.290   
30    JUN '16               NaN   1270.017       277.376  -238.040   -54.808   
31    MAR '16               NaN   1147.048       254.968  -248.224   -91.764   
32    DEC '15               NaN   1214.379       218.564  -260.333  -116.610   
33    SEP '15               NaN    936.789       231.496  -183.662   -73.296   
34    JUN '15               NaN    954.976       213.370  -170.188   -78.799   
35    MAR '15               NaN    939.880       260.073  -102.446   -25.334   
36    DEC '14               NaN    956.661       261.697   -74.838    -6.862   
37    SEP '14               NaN    851.804       251.851   -39.129    25.843   
38    JUN '14               NaN    769.349       212.995   -28.753    25.962   
39    MAR '14               NaN    620.542       155.128   -43.967     0.301   
40    DEC '13               NaN    615.219       156.589   -13.353    24.232   
41    SEP '13               NaN    431.346       102.868   -30.554    -2.104   
42    JUN '13               NaN    405.139       100.483   -11.792    10.406   
43    MAR '13               NaN    561.792        96.320    -5.584    12.266   
44    DEC '12               NaN    306.332        23.857   -90.883   -78.091   
45    SEP '12               NaN     50.104        -8.761  -108.460  -100.938   
46    JUN '12               NaN     26.653         4.762  -106.175  -101.827   
47    MAR '12               NaN     30.167        10.210   -88.763   -84.600   
48    DEC '11               NaN     39.375         7.835   -80.927   -76.123   
49 

We converts all Excel files in folder "Quarter Reports" to CSV files as starting from row 9 and column 2. Then saves the CSV files in an output folder   "Cleaned Quarter Reports".
As you can see some coulmns are empty let's drop them and keep the required columns only.

In [15]:
import os
import pandas as pd

def process_csv(folder_path):
    # List all files in the folder
    files = os.listdir(folder_path)
    
    for file in files:
        if file.endswith('.csv'):
            file_path = os.path.join(folder_path, file)
            
            # Read the CSV file
            df = pd.read_csv(file_path)
            
            # Extract the first column name (assuming it's the index or identifier column)
            first_column = df.columns[0]
            
            # Keep only the required columns
            required_columns = [
                '','Sales', 'Gross Income', 'EBIT', 'EBITDA', 'Net Income',
                'Cash & Short-Term Investments', 'Total Assets', 'Total Debt',
                'Net Debt', 'Total Liabilities', "Total Shareholders' Equity",
                'Net Operating Cash Flow', 'Capital Expenditures',
                'Net Investing Cash Flow', 'Net Financing Cash Flow',
                'Free Cash Flow', 
                'Long Term Debt Maturities'
            ]
            # Select required columns along with the first column
            columns_to_keep = [first_column] + [col for col in required_columns if col in df.columns]
            df = df[columns_to_keep]
            
            # Save the modified DataFrame back to CSV with the same filename
            df.to_csv(file_path, index=False)

# Provide the path to the folder containing CSV files
folder_path = 'Cleaned Quarter Reports'
process_csv(folder_path)

#Example output
df = pd.read_csv("Cleaned Quarter Reports/TESLA.csv")
df

Unnamed: 0      Sales  Gross Income      EBIT    EBITDA  Net Income  \
0     DEC '23  25167.000      4438.000  2064.000  3296.000    7927.000   
1     SEP '23  23350.000      4178.000  1764.000  2999.000    1851.000   
2     JUN '23  24927.000      4533.000  2399.000  3553.000    2703.000   
3     MAR '23  23329.000      4511.000  2664.000  3710.000    2518.000   
4     DEC '22  24318.000      5777.000  3935.000  4924.000    3722.000   
5     SEP '22  21454.000      5382.000  3688.000  4644.000    3292.000   
6     JUN '22  16934.000      4234.000  2606.000  3528.000    2256.000   
7     MAR '22  18756.000      5460.000  3603.000  4483.000    3313.000   
8     DEC '21  17719.000      4847.000  2613.000  3461.000    2321.000   
9     SEP '21  13757.000      3660.000  2055.000  2816.000    1618.000   
10    JUN '21  11958.000      2884.000  1335.000  2016.000    1142.000   
11    MAR '21  10389.000      2215.000   493.000  1114.000     438.000   
12    DEC '20  10744.000      2066.000   575.000  1193.000     270.000   
13    SEP '20   8771.000      2063.000   809.000  1393.000     300.000   
14    JUN '20   6036.000      1267.000   327.000   894.000     104.000   
15    MAR '20   5985.000      1234.000   283.000   836.000      16.000   
16    DEC '19   7384.000      1391.000   347.000   924.000     105.000   
17    SEP '19   6303.000      1191.000   261.000   791.851     143.000   
18    JUN '19   6349.676       921.046   -50.113   528.459    -408.334   
19    MAR '19   4541.464       565.743  -478.360   -10.783    -702.135   
20    DEC '18   7225.873      1442.900   419.151   915.888     139.483   
21    SEP '18   6824.413      1528.265   442.941   742.941     311.516   
22    JUN '18   4002.231       625.930  -517.958  -266.158    -717.539   
23    MAR '18   3408.751       462.326  -596.974  -180.741    -709.551   
24    DEC '17   3288.249       438.786  -598.141  -128.535    -675.350   
25    SEP '17   2984.675       459.940  -535.480  -134.856    -619.376   
26    JUN '17   2789.557       674.015  -240.916   -64.316    -336.397   
27    MAR '17   2696.270       674.046  -257.549   119.053    -330.277   
28    DEC '16   2284.631       374.878  -327.098    -0.159    -121.337   
29    SEP '16   2298.436       658.135    99.822   380.290      21.878   
30    JUN '16   1270.017       277.376  -238.040   -54.808    -293.188   
31    MAR '16   1147.048       254.968  -248.224   -91.764    -282.267   
32    DEC '15   1214.379       218.564  -260.333  -116.610    -320.397   
33    SEP '15    936.789       231.496  -183.662   -73.296    -229.858   
34    JUN '15    954.976       213.370  -170.188   -78.799    -184.227   
35    MAR '15    939.880       260.073  -102.446   -25.334    -154.181   
36    DEC '14    956.661       261.697   -74.838    -6.862    -107.629   
37    SEP '14    851.804       251.851   -39.129    25.843     -74.708   
38    JUN '14    769.349       212.995   -28.753    25.962     -61.900   
39    MAR '14    620.542       155.128   -43.967     0.301     -49.800   
40    DEC '13    615.219       156.589   -13.353    24.232     -16.264   
41    SEP '13    431.346       102.868   -30.554    -2.104     -38.496   
42    JUN '13    405.139       100.483   -11.792    10.406     -30.502   
43    MAR '13    561.792        96.320    -5.584    12.266      11.248   
44    DEC '12    306.332        23.857   -90.883   -78.091     -89.931   
45    SEP '12     50.104        -8.761  -108.460  -100.938    -110.804   
46    JUN '12     26.653         4.762  -106.175  -101.827    -105.603   
47    MAR '12     30.167        10.210   -88.763   -84.600     -89.873   
48    DEC '11     39.375         7.835   -80.927   -76.123     -81.489   
49    SEP '11     57.666        17.224   -64.477   -60.197     -65.078   
50    JUN '11     58.171        18.508   -58.739   -54.421     -58.903   
51    MAR '11     49.030        18.028   -47.346   -43.829     -48.941   
52    DEC '10     36.286        11.321   -51.645   -48.755     -51.358   
53    S

Now looks better. However , we need to rename first coulmn to "Date" and make sure it's a date type.

In [16]:
import os
import pandas as pd

# Specify the directory containing the CSV files
folder_path = 'Cleaned Quarter Reports'

# List all CSV files in the specified directory
csv_files = [file for file in os.listdir(folder_path) if file.endswith('.csv')]

# Iterate through each CSV file
for file_name in csv_files:
    file_path = os.path.join(folder_path, file_name)
    
    # Read the CSV file into a DataFrame
    df = pd.read_csv(file_path)
    
    # Check if the first column name is empty
    if df.columns[0] == 'Unnamed: 0':
        # Rename the first column to 'DATE'
        df.rename(columns={df.columns[0]: 'Date'}, inplace=True)
        
        # Save the modified DataFrame back to CSV
        df.to_csv(file_path, index=False)
        print(f"Renamed the first column in '{file_name}' to 'DATE'")
    else:
        print(f"No empty column found in '{file_name}'")

print("First Coulmn renamed")

import os
import pandas as pd

# Function to process each CSV file
def process_csv(file_path):
    # Read the CSV file
    df = pd.read_csv(file_path)
    
    # Convert the 'Date' column to datetime format
    df['Date'] = pd.to_datetime(df['Date'], format="%b '%y")
    
    # Convert the 'Date' column to the desired format
    df['Date'] = df['Date'].dt.strftime('%Y-%m-%d')
    
    # Save the updated data to the original CSV file
    df.to_csv(file_path, index=False)

# Folder containing CSV files
folder_path = 'Cleaned Quarter Reports'

# Iterate through each file in the folder
for filename in os.listdir(folder_path):
    if filename.endswith(".csv"):
        file_path = os.path.join(folder_path, filename)
        process_csv(file_path)



Renamed the first column in '1211.csv' to 'DATE'
Renamed the first column in '175.csv' to 'DATE'
Renamed the first column in '2333.csv' to 'DATE'
Renamed the first column in '500520.csv' to 'DATE'
Renamed the first column in '5380.csv' to 'DATE'
Renamed the first column in '7201.csv' to 'DATE'
Renamed the first column in '7202.csv' to 'DATE'
Renamed the first column in '7203.csv' to 'DATE'
Renamed the first column in '7261.csv' to 'DATE'
Renamed the first column in '7267.csv' to 'DATE'
Renamed the first column in '7269.csv' to 'DATE'
Renamed the first column in '7270.csv' to 'DATE'
Renamed the first column in '9868.csv' to 'DATE'
Renamed the first column in 'BMW.csv' to 'DATE'
Renamed the first column in 'Ford.csv' to 'DATE'
Renamed the first column in 'GM.csv' to 'DATE'
Renamed the first column in 'LCID.csv' to 'DATE'
Renamed the first column in 'LI.csv' to 'DATE'
Renamed the first column in 'MBG.csv' to 'DATE'
Renamed the first column in 'NIO.csv' to 'DATE'
Renamed the first column i

Now the dataset is ready , next step to combine it with historical stock prices

In [17]:
#Example output
df = pd.read_csv("Cleaned Quarter Reports/TESLA.csv")
df

Date      Sales  Gross Income      EBIT    EBITDA  Net Income  \
0   2023-12-01  25167.000      4438.000  2064.000  3296.000    7927.000   
1   2023-09-01  23350.000      4178.000  1764.000  2999.000    1851.000   
2   2023-06-01  24927.000      4533.000  2399.000  3553.000    2703.000   
3   2023-03-01  23329.000      4511.000  2664.000  3710.000    2518.000   
4   2022-12-01  24318.000      5777.000  3935.000  4924.000    3722.000   
5   2022-09-01  21454.000      5382.000  3688.000  4644.000    3292.000   
6   2022-06-01  16934.000      4234.000  2606.000  3528.000    2256.000   
7   2022-03-01  18756.000      5460.000  3603.000  4483.000    3313.000   
8   2021-12-01  17719.000      4847.000  2613.000  3461.000    2321.000   
9   2021-09-01  13757.000      3660.000  2055.000  2816.000    1618.000   
10  2021-06-01  11958.000      2884.000  1335.000  2016.000    1142.000   
11  2021-03-01  10389.000      2215.000   493.000  1114.000     438.000   
12  2020-12-01  10744.000      2066.000   575.000  1193.000     270.000   
13  2020-09-01   8771.000      2063.000   809.000  1393.000     300.000   
14  2020-06-01   6036.000      1267.000   327.000   894.000     104.000   
15  2020-03-01   5985.000      1234.000   283.000   836.000      16.000   
16  2019-12-01   7384.000      1391.000   347.000   924.000     105.000   
17  2019-09-01   6303.000      1191.000   261.000   791.851     143.000   
18  2019-06-01   6349.676       921.046   -50.113   528.459    -408.334   
19  2019-03-01   4541.464       565.743  -478.360   -10.783    -702.135   
20  2018-12-01   7225.873      1442.900   419.151   915.888     139.483   
21  2018-09-01   6824.413      1528.265   442.941   742.941     311.516   
22  2018-06-01   4002.231       625.930  -517.958  -266.158    -717.539   
23  2018-03-01   3408.751       462.326  -596.974  -180.741    -709.551   
24  2017-12-01   3288.249       438.786  -598.141  -128.535    -675.350   
25  2017-09-01   2984.675       459.940  -535.480  -134.856    -619.376   
26  2017-06-01   2789.557       674.015  -240.916   -64.316    -336.397   
27  2017-03-01   2696.270       674.046  -257.549   119.053    -330.277   
28  2016-12-01   2284.631       374.878  -327.098    -0.159    -121.337   
29  2016-09-01   2298.436       658.135    99.822   380.290      21.878   
30  2016-06-01   1270.017       277.376  -238.040   -54.808    -293.188   
31  2016-03-01   1147.048       254.968  -248.224   -91.764    -282.267   
32  2015-12-01   1214.379       218.564  -260.333  -116.610    -320.397   
33  2015-09-01    936.789       231.496  -183.662   -73.296    -229.858   
34  2015-06-01    954.976       213.370  -170.188   -78.799    -184.227   
35  2015-03-01    939.880       260.073  -102.446   -25.334    -154.181   
36  2014-12-01    956.661       261.697   -74.838    -6.862    -107.629   
37  2014-09-01    851.804       251.851   -39.129    25.843     -74.708   
38  2014-06-01    769.349       212.995   -28.753    25.962     -61.900   
39  2014-03-01    620.542       155.128   -43.967     0.301     -49.800   
40  2013-12-01    615.219       156.589   -13.353    24.232     -16.264   
41  2013-09-01    431.346       102.868   -30.554    -2.104     -38.496   
42  2013-06-01    405.139       100.483   -11.792    10.406     -30.502   
43  2013-03-01    561.792        96.320    -5.584    12.266      11.248   
44  2012-12-01    306.332        23.857   -90.883   -78.091     -89.931   
45  2012-09-01     50.104        -8.761  -108.460  -100.938    -110.804   
46  2012-06-01     26.653         4.762  -106.175  -101.827    -105.603   
47  2012-03-01     30.167        10.210   -88.763   -84.600     -89.873   
48  2011-12-01     39.375         7.835   -80.927   -76.123     -81.489   
49  2011-09-01     57.666        17.224   -64.477   -60.197     -65.078   
50  2011-06-01     58.171        18.508   -58.739   -54.421     -58.903   
51  2011-03-01     49.030        18.028   -47.346   -43.829     -48.941   
52  2010-12-01     36.286        11

# 2. Combining the data

this section for cleaning historical dataset and combining the historical dataset with quarter reports into one dataset

In [36]:
import pandas as pd

# Example TESLA
file_path = 'HistoricalData/TESLA.xlsx'

# Read the Excel file into a DataFrame
df = pd.read_excel(file_path)

# Print the first few rows of the DataFrame
print(df.head())


            Unnamed: 0 Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  \
0                  NaN        NaN        NaN        NaN        NaN        NaN   
1                 Date      Price       CVol     Change   % Change   % Return   
2  2024-03-15 00:00:00     163.57   97146830       1.07   0.658462   0.658462   
3  2024-03-14 00:00:00      162.5  126325695      -6.98   -4.11848   -4.11848   
4  2024-03-13 00:00:00     169.48  106524490      -8.06  -4.539822  -4.539822   

             Unnamed: 6           Unnamed: 7 Unnamed: 8 Unnamed: 9  \
0                   NaN                  NaN        NaN        NaN   
1  Total Return (Gross)  Cumulative Return %       Open       High   
2                163.57         10170.197165     163.16   165.1845   
3                 162.5          10103.01424     167.77     171.17   
4                169.48         10541.272944     173.05     176.05   

  Unnamed: 10          Unnamed: 11  
0         NaN                  NaN  
1         Low  Cum

This is example of historical stock prices dataset of Tesla

In [37]:
import os
import pandas as pd

def process_excel(input_folder, output_folder):
    # Create the output folder if it doesn't exist
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)
    
    # Get a list of Excel files in the input folder
    excel_files = [f for f in os.listdir(input_folder) if f.endswith('.xlsx')]
    
    for file_name in excel_files:
        # Read the Excel file, skipping the first two rows
        input_file_path = os.path.join(input_folder, file_name)
        df = pd.read_excel(input_file_path, skiprows=2)
        
        # Convert the "Date" column to datetime format and extract only the date part
        if 'Date' in df.columns:
            df['Date'] = pd.to_datetime(df['Date'], format='%m-%d-%y', errors='coerce').dt.date
        
        # Construct the output file path for CSV
        output_csv_file_path = os.path.join(output_folder, file_name.replace('.xlsx', '.csv'))
        
        # Save the modified DataFrame to a new CSV file with the same name
        df.to_csv(output_csv_file_path, index=False)
        print(f"Processed and saved as CSV: {file_name}")
        
        # Load the CSV file into a DataFrame for further processing
        df_csv = pd.read_csv(output_csv_file_path)
        print(f"Loaded CSV into DataFrame: {file_name.replace('.xlsx', '.csv')}")
        # Now you can use df_csv for further processing
        
# Specify the input and output folder names
input_folder = 'HistoricalData'
output_folder = 'HistoricalDataCleaned'

process_excel(input_folder, output_folder)


Processed and saved as CSV: 1211.xlsx
Loaded CSV into DataFrame: 1211.csv
Processed and saved as CSV: 175.xlsx
Loaded CSV into DataFrame: 175.csv
Processed and saved as CSV: 2333.xlsx
Loaded CSV into DataFrame: 2333.csv
Processed and saved as CSV: 500520.xlsx
Loaded CSV into DataFrame: 500520.csv
Processed and saved as CSV: 5380.xlsx
Loaded CSV into DataFrame: 5380.csv
Processed and saved as CSV: 7201.xlsx
Loaded CSV into DataFrame: 7201.csv
Processed and saved as CSV: 7202.xlsx
Loaded CSV into DataFrame: 7202.csv
Processed and saved as CSV: 7203.xlsx
Loaded CSV into DataFrame: 7203.csv
Processed and saved as CSV: 7261.xlsx
Loaded CSV into DataFrame: 7261.csv
Processed and saved as CSV: 7267.xlsx
Loaded CSV into DataFrame: 7267.csv
Processed and saved as CSV: 7269.xlsx
Loaded CSV into DataFrame: 7269.csv
Processed and saved as CSV: 7270.xlsx
Loaded CSV into DataFrame: 7270.csv
Processed and saved as CSV: 9868.xlsx
Loaded CSV into DataFrame: 9868.csv
Processed and saved as CSV: BMW.xlsx

Here we have connverted the historcal data csv into a dataframe and converting date into date type

In [38]:
import os
import pandas as pd

def add_filename_column(folder_path):
    # List all files in the folder
    files = os.listdir(folder_path)
    
    # Filter out only CSV files
    csv_files = [file for file in files if file.endswith('.csv')]
    
    # Iterate over each CSV file
    for file in csv_files:
        file_path = os.path.join(folder_path, file)
        
        # Read the CSV file
        df = pd.read_csv(file_path)
        
        # Get the file name without extension
        filename = os.path.splitext(file)[0]
        
        # Add a new column with the file name
        df['Stock'] = filename
        
        # Save the modified DataFrame back to CSV
        df.to_csv(file_path, index=False)

# Specify the folder path containing CSV files
folder_path = 'HistoricalDataCleaned'

# Call the function to add the filename column to each CSV file in the folder
add_filename_column(folder_path)
print("Done")


Done


Added the company name into the dataframe

In [39]:
import os
import pandas as pd

def process_csv_files(input_folder, output_folder):
    # Create the output folder if it doesn't exist
    if not os.path.exists(output_folder):
        os.makedirs(output_folder)

    # Iterate through each file in the input folder
    for filename in os.listdir(input_folder):
        if filename.endswith('.csv'):
            # Read the CSV file
            csv_path = os.path.join(input_folder, filename)
            df = pd.read_csv(csv_path)

            # Count NaN values in each column
            nan_counts = df.isna().sum()

            # Print the number of NaN values for each column
            print(f"NaN counts for {filename}:")
            print(nan_counts)

            # Drop rows with NaN values
            df = df.dropna()

            # Save the modified DataFrame to a new CSV file in the output folder
            output_path = os.path.join(output_folder, filename)
            df.to_csv(output_path, index=False)

            print(f"Processed {filename}, NaN rows dropped. Saved to {output_path}")

# Specify input and output folders
input_folder = 'HistoricalDataCleaned'
output_folder = 'CleanedHisPrices'

# Process CSV files in the input folder and save modified CSV files to the output folder
process_csv_files(input_folder, output_folder)


NaN counts for 1211.csv:
Date                     0
Price                    0
CVol                    10
Change                   5
% Change                 5
% Return                 5
Total Return (Gross)     0
Cumulative Return %      1
Open                     0
High                     0
Low                      0
Cumulative Change %      1
Stock                    0
dtype: int64
Processed 1211.csv, NaN rows dropped. Saved to CleanedHisPrices\1211.csv
NaN counts for 175.csv:
Date                     0
Price                    0
CVol                    15
Change                  30
% Change                30
% Return                30
Total Return (Gross)     0
Cumulative Return %      1
Open                     0
High                     0
Low                      0
Cumulative Change %      1
Stock                    0
dtype: int64
Processed 175.csv, NaN rows dropped. Saved to CleanedHisPrices\175.csv
NaN counts for 2333.csv:
Date                     0
Price                    0


Drop rows with NaN values as Some companies contain empty data points

In [46]:
import pandas as pd
import os

# Directories
cleaned_reports_directory = "Cleaned Quarter Reports"
historical_data_directory = "CleanedHisPrices"
output_directory = "Datasets"

# Create the output directory if it doesn't exist
if not os.path.exists(output_directory):
    os.makedirs(output_directory)

# Get list of common filenames in both directories
common_filenames = set(os.listdir(cleaned_reports_directory)) & set(os.listdir(historical_data_directory))

# Iterate over each common filename
for filename in common_filenames:
    print(f"Processed CSV: {filename}")
    # Check if the file is a CSV file
    if filename.endswith(".csv"):
        # Read CSV files from both directories
        cleaned_filepath = os.path.join(cleaned_reports_directory, filename)
        historical_filepath = os.path.join(historical_data_directory, filename)
        
        cleaned_df = pd.read_csv(cleaned_filepath)
        historical_df = pd.read_csv(historical_filepath)
        
        # Merge DataFrames based on "Date" column
        merged_df = pd.merge(cleaned_df, historical_df, on='Date')
        
        # Save merged DataFrame to a new CSV file with the same filename in the output directory
        output_filepath = os.path.join(output_directory, filename)
        merged_df.to_csv(output_filepath, index=False)


Processed CSV: Ford.csv
Processed CSV: PSNY.csv
Processed CSV: 500520.csv
Processed CSV: VFS.csv
Processed CSV: GM.csv
Processed CSV: LI.csv
Processed CSV: STLAM.csv
Processed CSV: 9868.csv
Processed CSV: 7270.csv
Processed CSV: RIVN.csv
Processed CSV: MBG.csv
Processed CSV: TESLA.csv
Processed CSV: 175.csv
Processed CSV: 7201.csv
Processed CSV: VOLVO.csv
Processed CSV: 7202.csv
Processed CSV: 7203.csv
Processed CSV: NIO.csv
Processed CSV: vw.csv
Processed CSV: 5380.csv
Processed CSV: RACE.csv
Processed CSV: RNO.csv
Processed CSV: LCID.csv
Processed CSV: 7267.csv
Processed CSV: 2333.csv
Processed CSV: BMW.csv
Processed CSV: 7261.csv
Processed CSV: 7269.csv
Processed CSV: PII.csv
Processed CSV: 1211.csv


-Here we merge the historcal dataset witht the quarter reports for each company

-Now the financial data and the stock prices are merged into one csv based on "Date" into folder "Datasets"

-Next we need to combine all companies into one csv

In [47]:
import pandas as pd
import os

# Assuming all your CSV files are stored in a folder named 'quarter_reports'
folder_path = 'Datasets'  # Replace 'Datasets' with the actual path to your folder

# Initialize an empty list to store DataFrames for each CSV file
dfs = []

try:
    # Iterate over each CSV file in the folder
    for file_name in os.listdir(folder_path):
        if file_name.endswith('.csv'):  # Check if the file is a CSV file
            file_path = os.path.join(folder_path, file_name)  # Get the full file path
            df = pd.read_csv(file_path)  # Read the CSV file into a DataFrame
            # Add a new column with the CSV file name (without .csv extension)
            df['Stock'] = os.path.splitext(file_name)[0]
            dfs.append(df)  # Append the DataFrame to the list
    if not dfs:
        raise FileNotFoundError("No CSV files found in the specified folder.")
        
    # Concatenate all DataFrames into a single DataFrame
    combined_df = pd.concat(dfs, ignore_index=True, sort=False)

    # Define the path to the "Combined" folder
    output_folder = 'Combined'

    # Create the output folder if it doesn't exist
    os.makedirs(output_folder, exist_ok=True)

    # Define the path to save the combined DataFrame
    output_file_path = os.path.join(output_folder, 'combined_data.csv')

    # Save the combined DataFrame to a CSV file
    combined_df.to_csv(output_file_path, index=False)

    print(f"Combined DataFrame saved to {output_file_path}")

    # Display the combined DataFrame
    print(combined_df)

except Exception as e:
    print("An error occurred:", e)


Combined DataFrame saved to Combined\combined_data.csv
            Date         Sales  Gross Income        EBIT      EBITDA  \
0     2023-06-01  19593.273781   3390.528084   821.72970         NaN   
1     2023-03-01  17235.981654   2809.461672   620.95953         NaN   
2     2022-12-01  21579.652932   3642.484188  1248.31461  2034.63484   
3     2022-09-01  16780.137875   2792.730051   950.21545  1766.71790   
4     2022-06-01  12477.423959   1526.700345   395.64160  1089.63737   
...          ...           ...           ...         ...         ...   
1191  2005-06-01  30920.932178   4139.814808   826.45346  3447.95880   
1192  2005-03-01  27350.308701   3381.921709   234.63720  2854.97111   
1193  2004-12-01  29315.240627   7077.701449  -957.77721  1451.57034   
1194  2004-09-01  25893.795148   3380.639922   260.33127  2831.86648   
1195  2004-06-01  28315.297610   3952.988505   684.12476  3220.68594   

       Net Income  Cash & Short-Term Investments   Total Assets    Total Debt  \

Now all companies are combined into one dataset with quarter reports and historical prices

In [59]:
import pandas as pd

file_path = 'Combined/combined_data.csv'

# Read the CSV file into a DataFrame
df = pd.read_csv(file_path)

# Display the DataFrame
df

Date         Sales  Gross Income        EBIT      EBITDA  \
0     2023-06-01  19593.273781   3390.528084   821.72970         NaN   
1     2023-03-01  17235.981654   2809.461672   620.95953         NaN   
2     2022-12-01  21579.652932   3642.484188  1248.31461  2034.63484   
3     2022-09-01  16780.137875   2792.730051   950.21545  1766.71790   
4     2022-06-01  12477.423959   1526.700345   395.64160  1089.63737   
...          ...           ...           ...         ...         ...   
1191  2005-06-01  30920.932178   4139.814808   826.45346  3447.95880   
1192  2005-03-01  27350.308701   3381.921709   234.63720  2854.97111   
1193  2004-12-01  29315.240627   7077.701449  -957.77721  1451.57034   
1194  2004-09-01  25893.795148   3380.639922   260.33127  2831.86648   
1195  2004-06-01  28315.297610   3952.988505   684.12476  3220.68594   

       Net Income  Cash & Short-Term Investments   Total Assets    Total Debt  \
0      972.077042                   12785.664554   81377.938919   4305.771373   
1      603.631767                    9708.379588   79615.096986   5594.476748   
2     1027.353496                   10371.303932   71390.052758   4061.847924   
3      834.082097                    6680.685390   60107.014766   4045.862724   
4      421.432430                    8430.373737   51821.924433   5604.617551   
...           ...                            ...            ...           ...   
1191   377.376008                   14668.227781  161062.371039  75773.333491   
1192    85.203454                   14961.568896  167319.519010  78311.700436   
1193   336.971686                   18862.308240  173444.340877  83268.996548   
1194    81.888241                   17421.533773  160290.033908           NaN   
1195   419.146862                   14017.026129  153089.869002           NaN   

          Net Debt  ...    Change  % Change  % Return  Total Return (Gross)  \
0     -8479.893181  ... -0.270395 -0.895398 -0.895400             32.742146   
1     -4113.902840  ...  1.682365  6.252711  6.252709             31.276810   
2     -6309.456008  ...  0.340681  1.388609  1.388604             27.213709   
3     -2634.822666  ... -1.221032 -3.956935 -3.956936             32.423855   
4     -2825.756187  ...  0.324284  0.908638  0.908642             39.383980   
...            ...  ...       ...       ...       ...                   ...   
1191           NaN  ...  0.691894  2.503557  2.503547             30.857887   
1192           NaN  ...  0.242832  0.773058  0.773059             32.998380   
1193  64406.688308  ...  0.384373  1.372370  1.372382             29.597525   
1194           NaN  ...  0.219983  0.970844  0.970862             23.850075   
1195           NaN  ... -0.115272 -0.457971 -0.457979             26.118387   

      Cumulative Return %       Open       High        Low  \
0             3723.834303  29.876852  30.489712  29.647030   
1             3552.703124  27.212618  28.715935  27.085218   
2             3078.188564  25.555984  26.121609  24.746110   
3             3686.662273  30.732770  30.783740  29.280228   
4             4499.509565  35.681973  36.370120  35.350636   
...                   ...        ...        ...        ...   
1191            19.408333  27.384401  28.401747  27.384401   
1192            27.691230  30.808943  31.778769  30.808943   
1193            14.531209  27.743698  28.506140  27.743698   
1194            -7.709245  22.754017  22.993422  22.754017   
1195             1.068263  25.096682  25.295284  24.709934   

      Cumulative Change %  Stock  
0             3395.171593   1211  
1             3238.749415   1211  
2             2805.019904   1211  
3             3361.194953   1211  
4             4105.860998   1211  
...                   ...    ...  
1191             9.619897     vw  
1192            22.491765     vw  
1193             9.867599     vw  
1194           -11.467242     vw  
1195            -3.047141     vw  

[1196 rows x 30 columns]

Full dataset